daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/SAXInfosetOutputter.scala

```diff
@@ -0,0 +1,266 @@
+      var prefixes: Seq[String] = Seq()
+      while( n != nsbEnd) {
+        val prefix = if(n.prefix == null) "" else n.prefix
+        prefixes :+= prefix
+        n = n.parent
+      }
+      prefixes.reverse.foreach(contentHandler.endPrefixMapping)
+      true
+    } catch {
+      case _: SAXException => false
+    }
+  }
+
+  /**
+   * Called by Daffodil internals to signify the beginning of a complex element.
+   *
+   * @param diComplex the complex element that is started. Various fields of
+   *                  DIComplex can be accessed to determine things like the
+   *                  nil, name, namespace, etc.
+   * @return true on sucess, false if there was an error and Daffodil should stop all
+   *         future calls to the InfosetOutputter
+   */
+  override def startComplex(diComplex: DIComplex): Boolean = {
+    val contentHandler = xmlReader.getContentHandler
+    val ns = if (diComplex.erd.thisElementsNamespace.isNoNamespace) "" else diComplex.erd.thisElementsNamespace.toString
+    val prefix = if(diComplex.erd.thisElementsNamespacePrefix == null) "" else  diComplex.erd.thisElementsNamespacePrefix
+    val elemName = diComplex.erd.name
+    val qName = if (prefix == "") elemName else s"$prefix:$elemName"
+
+    try {
+      val nsbStart = diComplex.erd.minimizedScope
+      val nsbEnd = if (diComplex.isRoot) scala.xml.TopScope else diComplex.diParent.erd.minimizedScope
+      var n = nsbStart
+      while( n != nsbEnd) {
+        val prefix = if(n.prefix == null) "" else n.prefix
+        val uri = if(n.uri == null) "" else n.uri
+        contentHandler.startPrefixMapping(prefix, uri)
+        n = n.parent
+      }
+
+      val attrs = if(isNilled(diComplex)) {
+        createNilAttribute()
+      } else {
+        new AttributesImpl()
+      }
+
+      contentHandler.startElement(ns, elemName, qName, attrs)
+      true
+    } catch {
+      case _: SAXException => false
+    }
+  }
+
+  /**
+   * Called by Daffodil internals to signify the end of a complex element.
+   *
+   * @param diComplex the complex element that is ended. Various fields of
+   *                  DIComplex can be accessed to determine things like the
+   *                  nil, name, namespace, etc.
+   * @return true on sucess, false if there was an error and Daffodil should stop all
+   *         future calls to the InfosetOutputter
+   */
+  override def endComplex(diComplex: DIComplex): Boolean = {
+    val contentHandler = xmlReader.getContentHandler
+    val ns = diComplex.erd.thisElementsNamespace
+    val nsb = diComplex.namedQName
+    try {
+      contentHandler.endElement(ns, nsb.local, nsb.toQNameString)
+
+      val nsbStart = diComplex.erd.minimizedScope
+      val nsbEnd = if (diComplex.isRoot) scala.xml.TopScope else diComplex.diParent.erd.minimizedScope
+      var n = nsbStart
+
+      // we store the the prefixes in the same order as we started them so we can
+      // reverse them and end them in the enclosing order
```

Comment:
Whyy do we want them in this specific order? Is this just to make the output more deterministic, or does it avoid outputting redundant prefix definitions? or...?

Reply:
So the initial intention was to ensure the order we started the mapping of the prefix matched the order we ended them, but as Steve mentioned in another comment, this doesn't matter as per the specs:

```Note that start/endPrefixMapping events are not guaranteed to be properly nested relative to each other: all startPrefixMapping events will occur immediately before the corresponding startElement event, and all endPrefixMapping events will occur immediately after the corresponding endElement event, but their order is not otherwise guaranteed.```

Reply:
That spec allows for non-determinstic behavior, even though it's intent is just not to over constrain implementations. 

I'm happy with any order, so long as it is deterministic - on the same data repeating a parse should result in the same order.

Reply:
Hmm, we use minimizedScope for it. But I don't think it's necessarily deterministic, would it be preferable to sort before we call the start/stopPrefix

Reply:
I think it's deterministic. I think the randomness you were seeing that we talked about was because JDOM, Scala etc when converted to text output namespaces prefixes in different orders. I'm pretty sure the SAX InfosetOutputer will always create events in the same order. They just might not necessarily match how JDOM/whatever ContentHandler stores it. But that's outside the scope of Daffodil. As long was we calling things in the same deterministic order, we should be fine.